### DISCUSSION TITLE
Language tools breaking changes

### DISCUSSION THREAD

**Discussion by JoviDeCroock** [vuejs/language-tools#4383] [Q&A] [answered]
Hey all,

Thank you for this project, we use it for the CLI of [gql.tada](https://github.com/0no-co/gql.tada/blob/main/packages/cli-utils/src/ts/transformers/vue.ts). Something that I noticed is that after 2.0.1 there were some breaking changes that we can partially resolve.

We were able to resolve the breaking change wehere `getDefaultVuePlugins` got renamed to `getBasePlugins`. However there is a bigger one, which requires some context on what we do.

Given the [following file](https://github.com/0no-co/gql.tada/blob/main/examples/example-pokemon-vue/src/components/PokemonList.vue) we will extract call-expressions like `graphql()` and analyze them for correctness, generate types, ... This seemingly changed as well as when we run the transform all of a sudden the `graphql()` invocations get lost in the transformed code, this did not happen before. This also feels like a breaking change, however I don't see how we can fix this on our end... Is there some compiler options we need to enable or is there some way to omit the plugin that does this?

**Comment by johnsoncodehk** [comment, marked as answer]
After a quick look, I figured this was the missing piece: https://github.com/vuejs/language-tools/blob/2cfa7ef4eb687b40a4dc1187a0e3c8f185a16823/packages/language-core/lib/languageModule.ts#L130-L134

`getBasePlugins` is not a use case intended for external use, you may want to use `createVueLanguagePlugin` instead.

```ts
const projectHost: TypeScriptProjectHost = {
	...ts.sys,
	configFileName: realTsConfig,
	getCurrentDirectory: () => rootPath,
	getProjectVersion: () => '0',
	getScriptFileNames: () => parsedCommandLine.fileNames,
	getCompilationSettings: () => parsedCommandLine.options,
	getScriptSnapshot,
	scriptIdToFileName: serviceEnv.typescript!.uriToFileName,
	fileNameToScriptId: serviceEnv.typescript!.fileNameToUri,
};
const vueLanguagePlugin = createVueLanguagePlugin(
	ts,
	serviceEnv.typescript!.uriToFileName,
	ts.sys.useCaseSensitiveFileNames,
	() => projectHost.getProjectVersion?.() ?? '',
	() => projectHost.getScriptFileNames(),
	parsedCommandLine.options,
	parsedCommandLine.vueOptions,
);
const language = createTypeScriptLanguage(
	ts,
	[vueLanguagePlugin],
	projectHost,
);
const sourceScript = language.scripts.set('.../test.vue', snapshot, 'vue');
if (sourceScript?.generated) {
	return sourceScript.generated.languagePlugin.typescript?.getServiceScript(sourceScript.generated.root)?.code;
}
```

Refs: https://github.com/vuejs/language-tools/blob/2cfa7ef4eb687b40a4dc1187a0e3c8f185a16823/packages/language-service/tests/utils/createTester.ts

**Reply by JoviDeCroock** [reply]
Thank you so much! As a quick fix we can push the `vueSfcPlugin` into the array, to start using the `createVueLanguagePlugin` we'll need to reconsider some architectural decisions especially when it comes to source maps
